feat: Configurable Reference Field

marination opened the pull request:
>[!NOTE]
> Text matches in unpaid vouchers: 
> - **Name in Desc**
> - **Reference in Desc** (New)
> - **Reference equals Bank Reference** (New)

- Set reference field in Banking Settings against a doctype
  <img width="700" alt="Screenshot 2024-11-20 at 6 23 36 PM" src="https://github.com/user-attachments/assets/5af52351-f82b-4ead-961d-0e129865c23b">
- Support for SI, PI and EC as these are the only documents to have an ambiguous reference field
- fix: don't manipulate reference field value if != "name" for INSTR as the field value will be ambiguous
-  <img width="700" alt="Screenshot 2024-11-20 at 6 19 45 PM" src="https://github.com/user-attachments/assets/13c857cc-d0c0-41d5-af70-5306dde212f9">
- WIth custom reference field
   <img width="700" alt="Screenshot 2024-12-11 at 4 56 06 PM" src="https://github.com/user-attachments/assets/1410b596-e306-4e25-ab20-3cc52af30233">

- Without custom reference field
   <img width="700" alt="Screenshot 2024-12-11 at 4 57 23 PM" src="https://github.com/user-attachments/assets/570e9d96-03c9-4bfd-b363-4a2ab12d0831">


- If a field is deleted and configuration still exists, that is indicated in the Tool
  <img width="700" alt="Screenshot 2024-11-21 at 5 15 11 PM" src="https://github.com/user-attachments/assets/a622bff5-f88d-430b-89f5-f4ea321072dc">


**Todo:**

- [x] Check for ambiguous column issues
- [x] Test
- [x] Abstract code

**To Test (Manual)**

- [x] Configure field in settings -> test if this field's value IN BT.description 
- [x] Configure field in settings -> test if this field's value == BT.reference_no
- [x] Remove configured field from settings -> test if voucher name appears in Reference No column
- [x] Test if voucher name IN BT.description

marination committed:
feat: Configurable Reference Field

- Set reference field in Banking Settings against a doctype
- Support for SI, PI and EC as these are the only documents to have an a,biguous reference field
- fix: don't manipulate reference field value if != "name" for INSTR as the field value will be ambiguous

marination committed:
fix: Add table alias in join queries to avoid ambiguous column errors for configurable reference column

marination committed:
test: `get_linked_payments` with configurable reference field

marination committed:
test: Case when ref field is not configured

- Also rollback to savepoint  after each test so that count of invoices assertion is unit tested

marination committed:
fix: Validate if the config points to a deleted field

marination committed:
fix(test): use setUp to make sure each test is completely atomic

- Tests might take longer, but this is the only sane way i can make this work
- Savepoints did not work for some reason and I'm running out of time :(

marination committed:
fix(test): Handle explicit commit in create_custom_field

- `create_custom_field` cannot be used arbitrarily in tests since it explicitly commits to make sure schema changes are reflected
- Call it during setUpClass since 2 tests need it
- Call it first so that after the commit, the following test setup can be rolled back on class cleanup
- Rollback to the post setup state after each test so that there is no test-run data shared between tests

barredterra reviewed (dismissed):
Custom reference should only supplement the existing one, not replace it.

marination commented:
Solution:
- Add another ranking "reference_in_desc" which considers reference no and make "name_in_desc" consider the voucher name
- Both should co-exist
- Since reference no field is introduced do a reference equality check if the ref_field != "name"

marination committed:
fix: Co-exist reference and name matches + reference equality match

- In some cases the BT contains some reference_field in its description, and sometimes the voucher name
- Check for both in BT desc. If both point to the "name" field then make reference check short to 0
- If reference field is configured, perform a reference equality check with BT reference_no

marination committed:
test: Adjust assertions as per new rankings

marination committed:
chore: extract out utils to reduce LOC in single file

marination committed:
fix: correct fieldname and misc

barredterra committed:
Merge branch 'version-15' into idm-244

barredterra commented on banking/klarna_kosma_integration/doctype/bank_reconciliation_tool_beta/bank_reconciliation_tool_beta.py:
```diff
@@ -1105,7 +1179,7 @@ def get_pi_matching_query(
 			purchase_invoice.name,
 			purchase_invoice.paid_amount,
-			purchase_invoice.bill_no.as_("reference_no"),
+			purchase_invoice[reference_field or "name"].as_("reference_no"),
```
Shouldn't this still default to `bill_no` instead of `name`? Might be an unpleasant change for the users, otherwise.

barredterra commented:
The functionality looks good, apart from these nitpicks. 👍🏼

barredterra commented on banking/klarna_kosma_integration/doctype/banking_settings/banking_settings.js:
```diff
@@ -256,6 +256,38 @@ frappe.ui.form.on('Banking Settings', {
+			let grid = frm.fields_dict.reference_fields.grid;
+			grid.update_docfield_property("field_name", "options", fieldnames);
+			frm.refresh_field("reference_fields");
+		});
```
This seems to be overindented

barredterra commented on banking/klarna_kosma_integration/doctype/banking_settings/banking_settings.json:
```diff
@@ -108,12 +111,29 @@
+   "label": "Bank Reconciliation"
+  },
+  {
+   "description": "<ul>\n<li>It stores DocType-wise mapping of fields that <b>should be considered as 'reference number' fields</b> in Bank Reconciliation Tool Beta</li>\n<li><b>Link</b> and <b>Data</b> fields are supported</li>\n<li>The field must have a <b>singular</b> reference number for matching with Bank Transactions</li></ul>",
```
Let's add a translation for this as well?

barredterra commented on banking/klarna_kosma_integration/doctype/bank_reconciliation_tool_beta/utils.py:
```diff
@@ -0,0 +1,93 @@
+# ref_in_desc_match: if ref in voucher IN bank statement description
+
+
+def amount_rank_condition(amount: any, bank_amount: float):
```
```suggestion
def amount_rank_condition(amount: Column, bank_amount: float) -> Case:
```

I think `any` (as in `any([False, True, False])`) is not a meaningful type.

marination committed:
fix: Over-indent + Typing + Translations

marination committed:
Merge branch 'version-15' into idm-244

marination commented:
@barredterra
- I've begrudgingly added the translations with HTML in the CSV as I'm not sure of any other way
- JS does not strip the the translatable string of its HTML tags and purely does a key-value store lookup. So this works for us.
- On the py side it would be problematic, but this particular translation is never used on the server side.
<img width="800" alt="Screenshot 2024-12-09 at 1 25 07 PM" src="https://github.com/user-attachments/assets/06d1ddb5-56cc-4c6b-95ff-515810f77206">

barredterra committed:
fix(Banking Settings): set field options on refresh (for existing rows)

barredterra committed:
feat(Banking Settings): show field label instead of name

barredterra committed:
fix: add missing import

barredterra committed:
fix: avoid increasing rank for `"" in "some_description"`

barredterra committed:
fix: german description

marination committed:
Merge branch 'version-15' into idm-244

marination committed:
fix: Move Reference column at the end + Saner column position handling

- Move Reference column at the end since it can be empty a lot of the times and should be a less important column (future: column order control)
- Its easy to forget where changes are to be done once the datatable column is reordered
- The column positions are used in multiple places to extract data
- Create a central mapping that stores and returns column positions, for sanity

marination commented:
@Mergifyio backport version-14
